src/platform/macos/events_loop/runloop_context.rs

```diff
@@ -0,0 +1,359 @@
+    INSIDE_INNER_RUNLOOP_CONTEXT.with(move |ctx| { ctx.set(context) });
+
+    // Yield back to `Runloop::new()` so it can return
+    // Our next execution -- and all subsequent executions -- will happen inside `Runloop::work()`.
+    yield_to_caller();
+
+    // Run the inner runloop
+    inner.run();
+
+    // Drop it
+    drop(inner);
+
+    // Yield forever
+    loop {
+        yield_to_caller();
+    }
+}
+
+
+// If we're inside the InnerRunloop, return the current Timeout.
+fn current_timeout() -> Option<Timeout> {
+    INSIDE_INNER_RUNLOOP_TIMEOUT.with(|timeout| {
+        timeout.get()
+    })
+}
+
+// If we're inside the InnerRunloop, context switch and return true;
+// if we're outside, do nothing and return false
+fn yield_to_caller() -> bool {
+    // See if we we're inside the inner runloop
+    // If we are in the inner runloop, take the context since we're leaving
+    if let Some(context) = INSIDE_INNER_RUNLOOP_CONTEXT.with(|context_cell| { context_cell.take() }) {
+        // Yield
+        let t = unsafe { context.resume(0) };
+        // We're returned
+
+        // t.context is the caller's context
+        let context = Some(t.context);
+        // t.data is a pointer to an Option<Timeout>
+        // take() it
+        let timeout: *mut Option<Timeout> = t.data as *mut Option<_>;
+        let timeout: Option<Timeout> =
+            unsafe { mem::transmute::<*mut Option<_>, &mut Option<_>>(timeout) }
+                .take();
+
+        // Does the caller want their thread back soon?
+        if timeout == Some(Timeout::Now) {
+            // Try to ensure we'll yield again soon, regardless of what happens inside Cocoa
+            guard_against_lengthy_operations();
+        }
+
+        // Store the new values in the thread local cells until we yield back
+        INSIDE_INNER_RUNLOOP_CONTEXT.with(move |context_cell| {
+            context_cell.set(context);
+        });
+        INSIDE_INNER_RUNLOOP_TIMEOUT.with(move |timeout_cell| {
+            timeout_cell.set(timeout);
+        });
+
+        true
+    } else {
+        false
+    }
+}
+
+
+fn guard_against_lengthy_operations() {
+    // Schedule a block to run in the near future, just in case
+    // We can get called repeatedly, and we only want the most recent call to matter, so keep track
+    // using an atomic counter
+    static INVOCATIONS: AtomicUsize = ATOMIC_USIZE_INIT;
+
+    // Get the current value of the counter, and increment it
+    let this_invocation = INVOCATIONS.fetch_add(1, Ordering::SeqCst);
+
+    // Queue a block in two milliseconds
+    dispatch::Queue::main().after_ms(2, move || {
+        // Get the most recent invocation, which is one before the current value of the counter
+        let current_counter = INVOCATIONS.load(Ordering::Acquire);
+        let (most_recent_invocation, _) = current_counter.overflowing_sub(1);
+
+        // Are we the most recent call?
+        if most_recent_invocation == this_invocation {
```

Comment:
Hmm. Thinking aloud in concrete terms:

Let's say we enqueue a block for 5 ms from now, return, and re-enter 2 ms later wanting another 5 ms timer. `dispatch_async_f()` does not provide a means to cancel a queued block, so at the moment we re-enter, our status is that we have a block queued for 3 ms from now for the previous cycle and we have no block queued for 5 ms for this cycle. 

The current code will queue another block and ignore any firings from the previous block(s). This wastes some RAM and some CPU cycles, but if you're polling events in a tight loop the way you described, it's not causing extra wakeups or context switches because you're never letting it sleep anyway. Does this overhead affect real programs? ¯\\\_(ツ)\_/¯

That said, we could reduce this overhead by scheduling blocks less precisely. This is a fallback mechanism to reduce worst-case latency, so there's not any kind of strict timeout policy, it just needs to get control reasonably soon even if the coroutine gets stuck somewhere inside a Cocoa resize handler.

We can waste fewer resources by keeping two bits of state: one indicating the current deadline (which we update unconditionally each time we enter with `Timeout::Now`), another indicating if there is presently a block queued. In this example, the second cycle would reset the deadline to 5 ms from now, but it would see there's a block queued and avoid scheduling another one. When the block gets invoked, it checks the deadline: deadlines in the past `yield_to_caller()`, and deadlines in the future prompt the block to schedule a followup. If the block in this example was invoked on time, it would schedule another for 2 ms from now; if it was invoked late, it'll yield immediately.

I'll hook that up and update the PR.